hedera-node/hedera-app/src/main/java/com/hedera/node/app/NewStateRoot.java

```diff
@@ -0,0 +1,711 @@
+            final ReadableSingletonState<T> instance = (ReadableSingletonState<T>) singletonInstances.get(stateKey);
+            if (instance != null) {
+                return instance;
+            }
+
+            final var md = stateMetadata.get(stateKey);
+            if (md == null || !md.stateDefinition().singleton()) {
+                throw new IllegalArgumentException("Unknown singleton state key '" + stateKey + "'");
+            }
+
+            final var ret = createReadableSingletonState(md);
+            singletonInstances.put(stateKey, ret);
+            return ret;
+        }
+
+        @NonNull
+        @Override
+        public <E> ReadableQueueState<E> getQueue(@NonNull String stateKey) {
+            final ReadableQueueState<E> instance = (ReadableQueueState<E>) queueInstances.get(stateKey);
+            if (instance != null) {
+                return instance;
+            }
+
+            final var md = stateMetadata.get(stateKey);
+            if (md == null || !md.stateDefinition().queue()) {
+                throw new IllegalArgumentException("Unknown queue state key '" + stateKey + "'");
+            }
+
+            final var ret = createReadableQueueState(md);
+            queueInstances.put(stateKey, ret);
+            return ret;
+        }
+
+        @Override
+        public boolean contains(@NonNull final String stateKey) {
+            return stateMetadata.containsKey(stateKey);
+        }
+
+        @NonNull
+        @Override
+        public Set<String> stateKeys() {
+            return stateKeys;
+        }
+
+        @NonNull
+        protected abstract ReadableKVState createReadableKVState(@NonNull StateMetadata md);
+
+        @NonNull
+        protected abstract ReadableSingletonState createReadableSingletonState(@NonNull StateMetadata md);
+
+        @NonNull
+        protected abstract ReadableQueueState createReadableQueueState(@NonNull StateMetadata md);
+
+        @NonNull
+        static String extractStateKey(@NonNull final StateMetadata<?, ?> md) {
+            return md.stateDefinition().stateKey();
+        }
+    }
+
+    /**
+     * An implementation of {@link ReadableStates} based on the merkle tree.
+     */
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    public final class MerkleReadableStates extends MerkleStates {
+
+        /**
+         * Create a new instance
+         *
+         * @param stateMetadata cannot be null
+         */
+        MerkleReadableStates(@NonNull final Map<String, StateMetadata<?, ?>> stateMetadata) {
+            super(stateMetadata);
+        }
+
+        @Override
+        @NonNull
+        protected ReadableKVState<?, ?> createReadableKVState(@NonNull final StateMetadata md) {
+            return new OnDiskReadableKVState<>(
+                    md.serviceName(),
+                    extractStateKey(md),
+                    Objects.requireNonNull(md.stateDefinition().keyCodec()),
```

Comment:
Maybe it makes sense to add helper methods to extract key and value codecs (by analogy to `extractStateKey`)